lib/js/app/queryCreator/components/Filters/Filters.tsx

```diff
@@ -0,0 +1,266 @@
+import React, { FC, useEffect, useRef, useMemo, useReducer } from 'react';
+import { useSelector } from 'react-redux';
+import shallowEqual from 'shallowequal';
+import { Button, Select, Input } from '@keen.io/ui-core';
+import { DatePicker } from './DatePicker';
+import { GeoCoordinates } from './GeoCoordinates';
+
+import {
+  addFilter,
+  removeFilter,
+  updateFilter,
+  resetFilters,
+  setFilters
+} from './actions';
+import { getCollectionSchema } from '../../modules/events';
+
+import { filtersReducer } from './reducer';
+
+import { DATA_TYPES, FILTER_OPERATORS, DEFAULT_TIMEFRAME_ABSOLUTE_VALUE } from './constants';
+import { convertFilters, getTypeFromValue, convertSchemaProp, convertDateToString, isFilterValid } from './utils';
+
+import { AppState, PropertyType, Operator, Filter } from '../../types';
+import { SchemaProp } from './types';
+
+import text from './text.json';
+
+type Props = {
+  /** Collection name */
+  collection: string;
+  /** Filters */
+  filters: Filter[];
+  /** Onchange handler */
+  onChange: (filters: Filter[]) => void;
+};
+
+const Filters: FC<Props> = ({ collection, filters, onChange }) => {
+  const collectionSchema = useSelector((state: AppState) =>
+    getCollectionSchema(state, collection)
+  );
+
+  const options = useMemo(() => {
+    if (collectionSchema) {
+      return Object.keys(collectionSchema).map((propertyName) => ({
+        label: propertyName,
+        value: propertyName,
+      }));
+    }
+
+    return [];
+  }, [collectionSchema]);
+
+  const dataTypes = Object.keys(DATA_TYPES).map(item => ({ label: DATA_TYPES[item], value: DATA_TYPES[item] }));
+
+  const getOperatorOptions = (type?: PropertyType) => {
+    if (!type) return FILTER_OPERATORS;
+
+    return FILTER_OPERATORS.filter(operator => operator.dataTypes.includes(type));
+  }
+
+  const [state, filtersDispatcher] = useReducer(filtersReducer, []);
+  const stateRef = useRef(state);
+
+  const filtersRef = useRef(filters);
+
+  useEffect(() => {
+    const localFilters = state.filter((v) => v !== null);
+    if (
+      !shallowEqual(filters, filtersRef.current) &&
+      !shallowEqual(filters, localFilters)
+    ) {
+      filtersDispatcher(setFilters(filters));
+    }
+    filtersRef.current = filters;
+  }, [filters])
+
+  useEffect(() => {
+    if (!shallowEqual(state, stateRef.current)) {
+      stateRef.current = state; console.log('use state');
+    }
+  }, [state])
+
+  const getInputValue = (filter: Filter) => {
+    if (filter?.propertyValue && typeof filter?.propertyValue === 'string') return filter.propertyValue;
+    return '';
+  };
+
+  const getSelectValue = (options: any, filter: Filter) => {
```

Comment:
Extract it to <FilterValue> component and add unit tests for specific UI component related with operator and property type.